services/runs.go

```diff
@@ -0,0 +1,315 @@
+package services
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/smartcontractkit/chainlink/adapters"
+	"github.com/smartcontractkit/chainlink/logger"
+	"github.com/smartcontractkit/chainlink/store"
+	"github.com/smartcontractkit/chainlink/store/assets"
+	"github.com/smartcontractkit/chainlink/store/models"
+	"github.com/smartcontractkit/chainlink/utils"
+)
+
+// ExecuteJob saves and immediately begins executing a run for a specified job
+// if it is ready.
+func ExecuteJob(
+	job models.JobSpec,
+	initiator models.Initiator,
+	input models.RunResult,
+	creationHeight *hexutil.Big,
+	store *store.Store) (*models.JobRun, error) {
+
+	logger.Debugw(fmt.Sprintf("New run triggered by %s", initiator.Type), []interface{}{
+		"job", job.ID,
+		"input_status", input.Status,
+		"creation_height", creationHeight.ToInt(),
+	}...)
+
+	run, err := NewRun(job, initiator, input, creationHeight, store)
+	if err != nil {
+		return nil, err
+	}
+
+	return run, saveAndTrigger(run, store)
+}
+
+// NewRun returns a run from an input job, in an initial state ready for
+// processing by the job runner system
+func NewRun(
+	job models.JobSpec,
+	initiator models.Initiator,
+	input models.RunResult,
+	currentHeight *hexutil.Big,
+	store *store.Store) (*models.JobRun, error) {
+
+	now := store.Clock.Now()
+	if !job.Started(now) {
+		return nil, RecurringScheduleJobError{
+			msg: fmt.Sprintf("Job runner: Job %v unstarted: %v before job's start time %v", job.ID, now, job.EndAt),
+		}
+	}
+
+	if job.Ended(now) {
+		return nil, RecurringScheduleJobError{
+			msg: fmt.Sprintf("Job runner: Job %v ended: %v past job's end time %v", job.ID, now, job.EndAt),
+		}
+	}
+
+	run := job.NewRun(initiator)
+
+	run.Overrides = input
+	run = run.ApplyResult(input)
+	run.CreationHeight = currentHeight
+	run.ObservedHeight = currentHeight
+
+	cost := assets.NewLink(0)
+	for i, taskRun := range run.TaskRuns {
+		adapter, err := adapters.For(taskRun.Task, store)
+
+		if err != nil {
+			run = run.ApplyResult(run.Result.WithError(err))
+			return &run, nil
+		}
+
+		mp := adapter.MinContractPayment()
+		cost.Add(cost, &mp)
+
+		if currentHeight != nil {
+			run.TaskRuns[i].MinimumConfirmations = utils.MaxUint64(
+				store.Config.MinIncomingConfirmations,
+				taskRun.Task.Confirmations,
+				adapter.MinConfs())
+		}
+	}
+
+	// input.Amount is always present for runs triggered by ethlogs
+	if input.Amount != nil {
```

Comment:
If needed, we can revisit the prospect of having initiator specific validations broken out in another PR.